dimagi/commcare-hq
# Prevent double charging

## Product Description


## Technical Summary

Ticket: https://dimagi-dev.atlassian.net/browse/SAAS-14924

Stripe allow us to provide [idempotency key](https://stripe.com/docs/api/idempotent_requests). 

> Keys are eligible to be removed from the system automatically after they're at least 24 hours old, and a new request is generated if a key is reused after the original has been pruned.

If we send the same key again in 24 hours, Stripe will return the result of the first request. And if the payment failed, we can still retry it the next day.

After discussing with Biyeun, tweak the autopay logic a little bit, to make sure if a charge failed or we attempt to double charge, no payment record should be created. We're risking not having the record for the attempt to charge clients. 
My biggest concern is my PR is kind of a revert of this [PR](https://github.com/dimagi/commcare-hq/pull/12513). The ticket in that PR is no longer accessible, so I don't know what will happen if we charge before create record.

## Safety Assurance

### Safety story


### Automated test coverage



Add new test `test_double_charge_is_prevented_and_only_one_payment_record_created`

### QA Plan


[QA-5697](https://dimagi-dev.atlassian.net/browse/QA-5657)

- **Autopay still work**: QA should trigger autopayments to make sure autopay still work as expected. Relavent payment record is created, email sent to customer regarding this payment.
- **Double charge is prevented**: QA should trigger autopayments on same invoice twice and make sure 1) the second attempt will error, 2) only one payment record is created, 3) we only sent out email about this payment to our customer once.

### Rollback instructions




### Labels & Review


[QA-5697]: https://dimagi-dev.atlassian.net/browse/QA-5697?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
